proof/hash_to_g2.saw

```diff
@@ -0,0 +1,452 @@
+          let hoist_ifs = run (for [
+              {{ \ c (x1:t_Fp_2) (x2:t_Fp_2) y1 y2 -> (if c then (x1,x2) else (y1,y2)).0 == if c then x1 else y1 }}
+            , {{ \ c (x1:t_Fp_2) (x2:t_Fp_2) y1 y2 -> (if c then (x1,x2) else (y1,y2)).1 == if c then x2 else y2 }}
+          ] (\e -> custom_prove_cryptol e w4));
+          simplify (addsimps hoist_ifs empty_ss);
+          let F_rewrites = run (for [ // unfolds F_prod and F_sum; note the use of pattern matching to produce "nice" saw-core terms
+              {{ \(x:[5]t_Fp_2) -> (F_prod Fp_2 x == Fp_2.mul(a, Fp_2.mul(b, Fp_2.mul(c, Fp_2.mul(d, e)))) where [a,b,c,d,e] = x) }}
+            , {{ \(x:[3]t_Fp_2) -> (F_prod Fp_2 x == Fp_2.mul(a, Fp_2.mul(b, c)) where [a,b,c] = x) }}
+            , {{ \(x:[3]t_Fp_2) -> (F_sum Fp_2 x == Fp_2.add(a, Fp_2.add(b, c)) where [a,b,c] = x)}}
+            ] (\ e -> custom_prove_cryptol e w4));
+          simplify (addsimps F_rewrites (cryptol_ss ())); // get rid of F_prod and F_sum
+          simplify (addsimps fp2_laws empty_ss); // try to normalize field expressions
+          simplify (add_cryptol_defs ["ecEq","PEqBit"] empty_ss); // transform those to boolEq?
+        // Finally:
+          let F_expt_rewrite = run (admit_cryptol {{ \x -> F_expt`{n=758} Fp_2 x `((p^^2-9)/16) == fp2_exp x `((p^^2-9)/16) }} );
+          simplify (addsimp F_expt_rewrite empty_ss);
+          w4_unint_z3 ["fp_rep", "fp_abs","sqrt_align","Fp_2","sqrt_fp2","is_square_fp2","fp2_exp","HashToCurveE2::sgn0","fp2_invariant","HE2::A'","HE2::B'","HE2::Z","ZZZ"];
+      })
+      (do {
+        unfolding ["select"];
+        simplify (addsimps [fix_vec_12_2, fix_vec_12] fp2_simpset);
+        simplify basic_ss; // NOTE: does not work without this...
+        w4_unint_z3 ["Fp_2","Fp","sqrt_align"];
+      }));
+  });
+
+// Next we prove that the implementation-level spec is equivalent to the abstract spec
+
+// Temporary assumptions:
+
+map_to_isogenous_E2_correct <- admit_cryptol {{ \u -> affinify HE2::Isogenous_Curve (map_to_isogenous_E2_impl u) == HE2::map_to_curve_simple_swu u }};
+
+// TODO
+map_to_isogenous_E2_thm <- really_custom_prove_cryptol {{ \u -> map_to_isogenous_E2_impl u == map_to_isogenous_E2_impl_2 u }}
+do {
+  assume_unsat;
+};
+
+isogeny_map_to_E2_correct <- admit_cryptol {{ \p -> affinify E' (isogeny_map_to_E2_impl p) == if is_point_projective HE2::Isogenous_Curve p then HE2::iso_map (affinify HE2::Isogenous_Curve p) else affinify E' (apply isogeny_map_to_E2_impl p) }};
+
+// We need to establish that we get a point on the isogenous curve
+map_to_isogenous_E2_on_curve <- custom_prove_cryptol {{ \u -> is_point_projective HE2::Isogenous_Curve (map_to_isogenous_E2_impl u) == True }}
+do {
+  simplify (addsimps [is_point_projective_lemma,swu_impl_xy_nz,map_to_isogenous_E2_correct,swu_on_curve] empty_ss);
+  w4_unint_z3 (concat fp_unints ["map_to_isogenous_E2_impl"]);
+};
+
+let aug_len = 48; // aug_len should be 96 or 48
+let msg_len = 32;
+let DST_len= 8;
+
```

Comment:
Do we need to run a proof for some other choices of sizes?  I'd guess not, that the sizes only affect the hash_to_field part.  But it still might be reassuring to see.

Reply:
Added another size